synapse/util/caches/lrucache.py

```diff
@@ -136,6 +169,28 @@ async def _expire_old_entries(clock: Clock, expiry_seconds: int) -> None:
         assert cache_entry is not None
         cache_entry.drop_from_cache()
 
+        # Check mem allocation periodically if we are evicting a bunch of caches
+        if jemalloc_interface and evicting_due_to_memory and (i + 1) % 100 == 0:
+            try:
+                jemalloc_interface.refresh_stats()
+                mem_usage = jemalloc_interface.get_stat("allocated")
+                if mem_usage < target_cache_memory_usage:
+                    evicting_due_to_memory = False
+                    logger.info("Stop memory-based cache eviction.")
+            except Exception:
+                logger.warning(
+                    "Unable to read allocated memory, this may affect memory-based cache eviction."
+                )
+                # If we've failed to read the current memory usage then we
+                # should stop trying to evict based on memory usage
+                evicting_due_to_memory = False
+
+            # pause a little to allow memory time to deallocate
+            if node.last_access_ts_secs > now - expiry_seconds:
+                await clock.sleep(0.5)
+            else:
+                await clock.sleep(0)
```

Comment:
Sorry, I meant updating the logic below where we already do a `await clock.sleep(0)` :slightly_smiling_face:

Reply:
Lol that's what I thought at first and then I second-guessed myself.

Reply:
I mean, its certainly a legitimate choice to put it there, and sleeping more often may actually be better in these circumstances :slightly_smiling_face: